asyncio_run_in_process/_child_trio.py

```diff
@@ -0,0 +1,135 @@
+)
+
+#
+# CLI invocation for subprocesses
+#
+parser = argparse.ArgumentParser(description="trio-run-in-process")
+parser.add_argument(
+    "--parent-pid", type=int, required=True, help="The PID of the parent process"
+)
+parser.add_argument(
+    "--fd-read",
+    type=int,
+    required=True,
+    help=(
+        "The file descriptor that the child process can use to read data that "
+        "has been written by the parent process"
+    ),
+)
+parser.add_argument(
+    "--fd-write",
+    type=int,
+    required=True,
+    help=(
+        "The file descriptor that the child process can use for writing data "
+        "meant to be read by the parent process"
+    ),
+)
+
+
+SHUTDOWN_SIGNALS = {signal.SIGTERM}
+
+
+async def _do_monitor_signals(signal_aiter: AsyncIterator[int]) -> None:
+    async for signum in signal_aiter:
+        raise SystemExit(signum)
+
+
+@trio_typing.takes_callable_and_args
+async def _do_async_fn(
+    async_fn: Callable[..., Awaitable[TReturn]],
+    args: Sequence[Any],
+    to_parent: BinaryIO,
+) -> TReturn:
+    with trio.open_signal_receiver(*SHUTDOWN_SIGNALS) as signal_aiter:
+        # state: STARTED
+        update_state(to_parent, State.STARTED)
+
+        async with trio.open_nursery() as nursery:
+            nursery.start_soon(_do_monitor_signals, signal_aiter)
+
+            # state: EXECUTING
+            update_state(to_parent, State.EXECUTING)
+
+            result = await async_fn(*args)
+
+            # state: STOPPING
+            update_state(to_parent, State.STOPPING)
+
+            nursery.cancel_scope.cancel()
+        return result
+
+
+def _run_process(parent_pid: int, fd_read: int, fd_write: int) -> None:
+    """
+    Run the child process
+    """
+    # state: INITIALIZING
+    with os.fdopen(fd_write, "wb", closefd=True) as to_parent:
+        # state: INITIALIZED
+        update_state_initialized(to_parent)
+        with os.fdopen(fd_read, "rb", closefd=True) as from_parent:
+            # state: WAIT_EXEC_DATA
+            update_state(to_parent, State.WAIT_EXEC_DATA)
+            async_fn, args = receive_pickled_value(from_parent)
+
+        # state: BOOTING
+        update_state(to_parent, State.BOOTING)
+
+        try:
+            try:
+                result = trio.run(_do_async_fn, async_fn, args, to_parent)
+            except BaseException as err:
+                # state: STOPPING
+                update_state(to_parent, State.STOPPING)
+                finished_payload = pickle_value(err)
+                raise
+        except KeyboardInterrupt:
+            code = 2
+        except SystemExit as err:
+            code = err.args[0]
+        except BaseException:
+            code = 1
+        else:
+            # state: STOPPING (set from within _do_async_fn)
+            finished_payload = pickle_value(result)
+            code = 0
+        finally:
+            # state: FINISHED
+            update_state_finished(to_parent, finished_payload)
+            sys.exit(code)
+
```

Comment:
It looks a lot like we could pretty easily DRY this with [the `asyncio` version](https://github.com/ethereum/asyncio-run-in-process/blob/9ff64a1c6bc27fb8a4be27d9a2a2b97003b4c627/asyncio_run_in_process/_child.py#L167-L218)

The only place they differ is 1) in invoking the `trio.run` vs `loop.run_unitl_complete` . and 2) in the exception handling which special cases `asyncio.CancelledError`

Both of those should be reasonably trivial to abstract into something generic we can use for both.